Rework Python packaging lecture (#105)
Simulation-Software-Engineering/Lecture-Material: 03_building_and_packaging/pypi_slides.md

---

## Python packaging is itself evolving

Files which are commonly seen:

`setup.py`, `setup.cfg`, `pyproject.toml`, `requirements.txt`.

> All are files which packaging-related tools consume. What do each of these files do?

Reviewer: ```suggestion
All are files which packaging-related tools consume. What does each of these files do?
```
or 

```suggestion
All are files which packaging-related tools consume. What do these files do?
```

---

## Python Enhancement Proposals (PEPs)

- PEP is an evolving design document which provides information regarding new features of Python, new processes and new environments.
- PEPs typically involve concise technical information, which also acts as standardizations.
- Packaging workflows are also standardized through PEPs. Examples are
    - [PEP 427](https://www.python.org/dev/peps/pep-0427/) which introduces the built-package format "wheel".
    - [PEP 518](https://peps.python.org/pep-0518/) which introduces a configuration file for packaging.
- Tip: read the *Rationale* section of the **PEP** convention of a particular feature.

---

## Python libraries used to install packages

- `disutils`: old and deprecated, to be phased out in Python 3.12.
- `setuptools`: actively maintained packaging tool which is shipped with Python (built on top of `disutils`).

---

## setup.py - setup.cfg - pyproject.toml

- Names of all these files are standardized.

> - `setup.py` is the interface to the command line interface for packaging. Needs to be at the root of the repository.

Reviewer: interface to the interface?

> - `setup.cfg` is a having metadata of all the options that are specified in `setup.py`.

Reviewer: Sth is wrong here (language)

    package_dir={"": "<directory-name>"},
    packages=setuptools.find_packages(where="<directory-name>"),
    python_requires=">=3.6",
    install_requires=["<dependencies>"]
)
```

---

## Using setup.cfg and setup.py

Entries moved to `setup.cfg` would look like:

```python
[metadata]
name="package-name"
version="<version-number>"
author="Your Name"
url="package-website-url"

[options]
install_requires =
  "<dependencies>"
```

Additionally `setup.cfg` can have flags for packages, for example `[bdist_wheel]`.

---

## Using pyproject.toml

- According to PEP 621, using `pyproject.toml` is the default recommended way of creating packages with `setuptools`.
- Most important table is `[build-system]` which specifies minimum requirements of the package (PEP 518).
- `pyproject.toml` is readable by packaging tools like pip.

Example `pyproject.toml` can look like

```python
[build-system]
requires = ["setuptools", "wheel"]
```

---

> ## Packaging tool build

Reviewer: ```suggestion
## Packaging tool `build`
```

**Important**: Do not use

```bash
sudo pip install <package>
```

Various security issues with doing so! Go for

```bash

> pip install --user <package>

Reviewer: Ideas for extension: Explain how to uninstall and how to update a package

---

## File structure for code packaging 1/2

Three naming points:

- Name of the repository (on GitHub or GitLab).
- Name of the folder which has the source code.
- Name of the package as seen my PyPI.
- Name of the package to be used in the `import` statement.

> **All three names are independent of each other.**

Reviewer: What is a "naming point"?
And I count four points :grin:

Example folder structure:

```bash
generic_folder_name/
└── src/
     ├── __init__.py
     └── source-code.py
```

- The file `__init__.py` is required to import the `package_name/` package as a package. This file is mostly empty

> - `source-code.py` will contain the code. It can be multiple files.

Reviewer: ```suggestion
- `source-code.py` contains the code. It can be multiple files.
```
